Issue #419 - Move URL computation to ClearlyDefined summarizer

storrisi opened the pull request:
Fixes #419

jeffmcaffer commented on business/definitionService.js:
```diff
@@ -504,10 +504,10 @@ class DefinitionService {
     }
   }
 
-  async _ensureUrls(coordinates, definition) {
```
push this level of function down to clearlydefined.js or the appropriate summarizer. Should not need to access the raw results after the `summarizeAll` call.

storrisi commented:
@jeffmcaffer i've updated the PR moving the URL building methods into the summarizer.
I've also added URL supporting for both `pod` and `sourcearchive`, which fixes #426 

For pod i didn't found a method to retrieve a specific download url in the harvested data, looks like everything is based on github. Is that right?

jeffmcaffer commented on providers/summary/clearlydefined.js:
```diff
@@ -132,20 +135,35 @@ class ClearlyDescribedSummarizer {
     let licenses = licenseUrls.map(extractLicenseFromLicenseUrl).filter(x => x)
     if (!licenses.length) licenses = licenseNames.map(x => SPDX.lookupByName(x) || x).filter(x => x)
     if (licenses.length) setIfValue(result, 'licensed.declared', SPDX.normalize(licenses.join(' OR ')))
+    const registryUrl = buildRegistryUrl(coordinates)
```
This code is replicated many times. Break it out into a function that is shared. Something like `addStandardUrls` perhaps

jeffmcaffer commented on providers/summary/clearlydefined.js:
```diff
@@ -132,20 +135,35 @@ class ClearlyDescribedSummarizer {
+    const versionUrl = buildVersionUrl(coordinates)
+    const downloadUrl = buildDownloadUrl(coordinates)
+    if (get(result, 'described'))
+      setIfValue(result, 'described.urls', { registry: registryUrl, version: versionUrl, download: downloadUrl })
```
If the consts happened to have matching names (e.g., `registry` rather than `registryUrl`) then this could just be `{ registry, version, download }`

storrisi replied:
Right, we can save some code doing this

jeffmcaffer replied:
mostly just cleaner. The real code saving comes from removing all the duplicate code.

jeffmcaffer commented on providers/summary/clearlydefined.js:
```diff
@@ -132,20 +135,35 @@ class ClearlyDescribedSummarizer {
+    const registryUrl = buildRegistryUrl(coordinates)
+    const versionUrl = buildVersionUrl(coordinates)
+    const downloadUrl = buildDownloadUrl(coordinates)
+    if (get(result, 'described'))
```
The if is not needed. We want to set these values regardless of whether there is a `described` prop. Setting will create the `described` prop

storrisi replied:
I've changed this after updating tests. 
Looks like that in some cases the described prop should be undefined at all, like `expect(summary.described).to.be.undefined`

jeffmcaffer replied:
We were likely expecting undefined in the tests because previously there was nothing being added to `described` so there should not have been one. Now we have changed the system to add in URLs so the expectation should change.

jeffmcaffer commented on providers/summary/clearlydefined.js:
```diff
@@ -179,30 +202,58 @@ class ClearlyDescribedSummarizer {
+    const releases = get(data, 'registryData.releases')
+    let downloadUrl
+    forEach(releases[coordinates.revision], revision => {
+      if (revision.filename.includes('tar.gz')) downloadUrl = revision.url
```
Do you want the first or last one?

storrisi replied:
That's a good question.
Looks like that Pypi could have multiple objects for the same release.

For example:

```
"0.7": [
            {
              "comment_text": "",
              "digests": {
                "md5": "235729aa2bf99e1b940eba53fcbfb37c",
                "sha256": "bd053a6d0e5231bec41da6ef524369d64dc9a7f27c5075914ae1b1abc4dd33e5"
              },
              "downloads": -1,
              "filename": "idna-0.7.tar.gz",
              "has_sig": false,
              "md5_digest": "235729aa2bf99e1b940eba53fcbfb37c",
              "packagetype": "sdist",
              "python_version": "source",
              "size": 20294,
              "upload_time": "2014-07-10T03:37:29",
              "url": "https://files.pythonhosted.org/packages/cf/57/53ac056e1a9d028dfce896929d0fc769e53a1e4d0917c2ec163acae4528f/idna-0.7.tar.gz"
            }
          ],
```

has only one object that contains the .tar.gz file URL, in other cases i found multiple objects for the same release, with different file URL and extension.

What do you think could be the best way in this case?

jeffmcaffer replied:
No idea unfortunately of how to distinguish between multiple entries for the same component version. Propose the following
* add a TODO comment here noting that we are randomly picking the last (or first) entry that has a URL.
* Open an issue to figure out what the right answer is and in that issue, put some examples of components that have multiple entries for the same version with different URLs

From there we can get some pypi experts to weigh in on the situation and give us advice.

storrisi replied:
Thanks, i will follow your suggestions

jeffmcaffer commented on business/definitionService.js:
```diff
@@ -261,9 +258,9 @@ class DefinitionService {
   // Compute and store the scored for the given definition but do it in a way that does not affect the
   // definition so that further curations can be done.
-  async _ensureToolScores(coordinates, definition) {
+  async _ensureToolScores(coordinates, definition, raw) {
```
this whole call chain no longer needs async

storrisi replied:
Actually this method could be removed at all. I will fix it. Thanks!

jeffmcaffer replied:
not sure what you mean about removing methods. All the methods here were here before and need to stay. Just the async/awaits no longer need to be added.

storrisi replied:
Ah, sorry, i thought you were referring to the `ensureUrls` method that actually is not used anymore. My bad.

storrisi commented:
@jeffmcaffer i've updated the PR based on your suggestions, as well as tests.

jeffmcaffer commented on lib/utils.js:
```diff
@@ -239,12 +247,14 @@ function buildVersionUrl(spec) {
 }
 
-async function buildDownloadUrl(spec) {
+function buildDownloadUrl(spec) {
```
might be good to add a comment here that this will build URLs for any protocol with URLs that can be predicted from the identity of the component

jeffmcaffer commented on test/summary/clearlydefinedTests.js:
```diff
@@ -140,17 +146,55 @@ describe('ClearlyDescribedSummarizer addNpmData', () => {
+      summarizer.addNpmData(result, { registryData: { manifest: { license: { type: license } } } }, npmTestCoordinates)
+      if (data[license])
+        assert.deepEqual(result, {
+          described: {
```
this duplicate code can be collapsed by conditionally creating the target opbject and then asserting. So 

```js
const expected = {
  described: {
    urls: {
      download: 'https://registry.npmjs.com/test/-/test-1.0.tgz',
      registry: 'https://npmjs.com/package/test',
      version: 'https://npmjs.com/package/test/v/1.0'
  }
}
if (data[license]) expected.license = { declared: data[license] }
assert.deepEqual(result, expected)
```

Since the `urls` part is reused so much thoughout the tests, perhaps that should be a constant somewhere as well.

jeffmcaffer commented on test/summary/clearlydefinedTests.js:
```diff
@@ -166,51 +210,135 @@ describe('ClearlyDescribedSummarizer addNpmData', () => {
+      assert.deepEqual(result, {
+        described: {
+          releaseDate: data[date],
+          urls: {
```
make a reusable `urls` prop so we don't have to duplicate this everywhere

jeffmcaffer commented on test/summary/clearlydefinedTests.js:
```diff
@@ -226,9 +354,28 @@ describe('ClearlyDescribedSummarizer addNuGetData', () => {
-      else assert.deepEqual(result, {})
+      summarizer.addNuGetData(result, { manifest: { licenseExpression } }, nugetTestCoordinates)
+      if (data[licenseExpression])
+        assert.deepEqual(result, {
```
see above comments on duplicate code elimination

storrisi commented:
Thanks @jeffmcaffer , i've updated the PR and you can review it again.

dabutvin commented on providers/summary/clearlydefined.js:
```diff
@@ -181,28 +189,44 @@ class ClearlyDescribedSummarizer {
+
+    const registry = buildRegistryUrl(coordinates)
+    const version = homepage && revision ? `${homepage}/tree/${revision}` : null
+    const download = homepage && revision ? `${homepage}/archive/${revision}.zip` : null
```
This only works when the homepage is on github

This one for example would not work, https://api.clearlydefined.io/definitions/pod/cocoapods/-/crashlytics/3.12.0

dabutvin replied:
take a look at https://github.com/clearlydefined/crawler/blob/master/providers/fetch/podFetch.js#L75 for how we resolve it.

storrisi replied:
Ah, thanks, i hadn't any example for that case.
Looking at the harvested data of the example, looks like that we can't link to a specific version URL, since it doesn't exists.

dabutvin replied:
that's what I found too :( that's okay

dabutvin commented on providers/summary/clearlydefined.js:
```diff
@@ -181,28 +189,44 @@ class ClearlyDescribedSummarizer {
+    const revision = find(releases[coordinates.revision], revision => revision.filename.includes('tar.gz'))
+    const registry = buildRegistryUrl(coordinates)
+    const version = buildVersionUrl(coordinates)
+    setIfValue(result, 'described.urls', { registry, version, download: revision.url })
```
setIfValue doesn't do much here since we are providing an object for the value inline. It will never be falsy and not set.

You may want to setIfValue each individual property

storrisi replied:
Right, thanks

dabutvin commented on providers/summary/clearlydefined.js:
```diff
@@ -11,7 +11,10 @@ const {
+  buildUrls,
+  buildRegistryUrl,
+  buildVersionUrl
 } = require('../../lib/utils')
```
I think we should probably move `buildUrls` out of utils now that it won't handle all coordinates thrown at it.
Probably better just to put each in it's respective spot here.

so for example `addNpmData` will add the urls for npms

storrisi replied:
I'm not 100% sure about this.
Most of the providers use the same path to retrieve the URLs.
Only in some specific cases (Pypi, Cocoapods) we must use different ways.

Removing that function would potentially cause some code duplication.

Thoughts?

dabutvin replied:
I don't think code duplication will be a problem here

It would be slotting these into each of the methods instead of living in a big partial switch

```js
function buildRegistryUrl(spec) {
  switch (spec.provider) {
    case 'github':
      return `https://github.com/${spec.namespace}/${spec.name}`
    case 'npmjs':
      return `https://npmjs.com/package/${spec.namespace ? spec.namespace + '/' + spec.name : spec.name}`
    case 'nuget':
      return `https://nuget.org/packages/${spec.name}`
    case 'cratesio':
      return `https://crates.io/crates/${spec.name}`
    case 'mavencentral':
      return `https://mvnrepository.com/artifact/${spec.namespace}/${spec.name}`
    case 'pypi':
      return `https://pypi.org/project/${spec.name}`
    case 'rubygems':
      return `https://rubygems.org/gems/${spec.name}`
    case 'cocoapods':
      return `https://cocoapods.org/pods/${spec.name}`
    default:
      return null
  }
```

storrisi replied:
Right, will address this soon.

storrisi replied:
Maybe i'm wrong but i can't found the specific method for `github` in the summarizer. Is that right?

dabutvin replied:
you're right. We'll have to add one - good catch

storrisi commented:
@dabutvin i've refactored the methods following your suggestions.
I've also added the support for `git` type.

dabutvin commented:
👍
